src/mdns/mdns_avahi.cpp

```diff
@@ -860,6 +919,262 @@ void Publisher::Destroy(Publisher *aPublisher)
     delete static_cast<PublisherAvahi *>(aPublisher);
 }
 
+void PublisherAvahi::ServiceSubscription::Browse(void)
+{
+    assert(mPublisherAvahi->mClient != nullptr);
+
+    otbrLogInfo("browse service %s", mType.c_str());
+    if (!(mServiceBrowser = avahi_service_browser_new(mPublisherAvahi->mClient, AVAHI_IF_UNSPEC, AVAHI_PROTO_INET6,
+                                                      mType.c_str(), mPublisherAvahi->mDomain,
+                                                      static_cast<AvahiLookupFlags>(0), HandleBrowseResult, this)))
+    {
+        otbrLogWarning("failed to browse service %s: %s", mType.c_str(),
+                       avahi_strerror(avahi_client_errno(mPublisherAvahi->mClient)));
+    }
+}
+
+void PublisherAvahi::ServiceSubscription::Release(void)
+{
+    if (mServiceBrowser != nullptr)
+    {
+        avahi_service_browser_free(mServiceBrowser);
+        mServiceBrowser = nullptr;
+    }
+    if (mServiceResolver != nullptr)
+    {
+        avahi_service_resolver_free(mServiceResolver);
+        mServiceResolver = nullptr;
+    }
+}
+
+void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser *  aServiceBrowser,
+                                                             AvahiIfIndex           aInterfaceIndex,
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags,
+                                                             void *                 aContext)
+{
+    static_cast<PublisherAvahi::ServiceSubscription *>(aContext)->HandleBrowseResult(
+        aServiceBrowser, aInterfaceIndex, aProtocol, aEvent, aName, aType, aDomain, aFlags);
+}
+
+void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser *  aServiceBrowser,
+                                                             AvahiIfIndex           aInterfaceIndex,
+                                                             AvahiProtocol          aProtocol,
+                                                             AvahiBrowserEvent      aEvent,
+                                                             const char *           aName,
+                                                             const char *           aType,
+                                                             const char *           aDomain,
+                                                             AvahiLookupResultFlags aFlags)
+{
+    OTBR_UNUSED_VARIABLE(aProtocol);
+
+    VerifyOrExit(mServiceBrowser == aServiceBrowser);
```

Comment:
use assert?

Reply:
It's not used for catching a logical error. I used `VerifyOrExit` because it may be possible (or may not) that aServiceBrowser has been freed before triggering this callback, at that time mServiceBrowser has been set to `nullptr` and we shouldn't do any work in the callback.

Reply:
If `mServiceBrowser` has been freed, will this callback still be triggered?